src/http.cc

```diff
@@ -1589,29 +1596,45 @@ HttpStateData::mayReadVirginReplyBody() const
 void
 HttpStateData::maybeReadVirginBody()
 {
-    // too late to read
-    if (!Comm::IsConnOpen(serverConnection) || fd_table[serverConnection->fd].closing())
+    if (!Comm::IsConnOpen(serverConnection) || fd_table[serverConnection->fd].closing()) {
+        debugs(11, 3, "no, peer connection gone");
         return;
+    }
+
+    if (eof) {
+        // tolerate hypothetical calls between Comm::ENDFILE and closeServer()
+        debugs(11, 3, "no, saw EOF");
+        return;
+    }
+
+    if (lastChunk) {
+        // tolerate hypothetical calls between setting lastChunk and clearing serverConnection
+        debugs(11, 3, "no, saw last-chunk");
+        return;
+    }
 
     if (!canBufferMoreReplyBytes()) {
         abortTransaction("more response bytes required, but the read buffer is full and cannot be drained");
         return;
     }
 
-    // XXX: get rid of the do_next_read flag
-    // check for the proper reasons preventing read(2)
-    if (!flags.do_next_read)
+    if (waitingForDelayAwareReadChance) {
```

Comment:
Should not we add a similar check into Ftp::Client::maybeReadVirginBody() or assert  !waitingForDelayAwareReadChance there?

Reply:
> Should not we add a similar check into Ftp::Client::maybeReadVirginBody() 

I would rather not because Ftp::Client::maybeReadVirginBody() already has code that prevents "duplicate" delayRead() calls: Ftp::Client uses a similar-but-different FTP-specific flag[^1] for that purpose. This PR already has an Ftp::Client TODO comment to address that code duplication and its side effects. There are quite a few things that should be refreshed in that FTP code, to bring them in sync with HTTP (and then move some of them into Client)!

Said that, I will add this check if others insist.

[^1]: Ftp::Client uses data.read_pending. AFAICT, that DataChannel flag combines waitingForDelayAwareReadChance and waitingForCommRead logic. The flags.do_next_read bug fixed in this PR is a good illustration why such combinations should be avoided.

> or assert !waitingForDelayAwareReadChance there?

I hesitate adding that assertion into that poorly understood, poorly duplicated, and rather neglected FTP code that, according to our tests (thank you for running them!), does not suffer from Bug 5352. I will add either this assertion or the check discussed above if others insist.